### ISSUE TITLE
Elementwise hardlogistic

### ISSUE THREAD

**Issue by zamir1001** [oneapi-src/oneDNN#1199]
# Summary
The addition of a dnnl_eltwise_hardlogistic activation function

# Problem statement
There's already an activation function dnnl_eltwise_swish and dnnl_eltwise_hardswish.
There's also an activation function dnnl_elementwise_logistic but not a dnnl_elementwise_hardlogistic.
[wikipedia hardsigmoid](https://en.wikipedia.org/wiki/Hard_sigmoid)

# Preferred solution
y = max(0, min(1, alpha * x + beta))

dy/x =  alpha  (0 < alpha * x + beta < 1)
dy/x =  0  (otherwise)

[pytorch hardsigmoid](https://pytorch.org/docs/stable/generated/torch.nn.Hardsigmoid.html)
[tensorflow hardsigmoid](https://www.tensorflow.org/api_docs/python/tf/keras/activations/hard_sigmoid)
[mxnet hardsigmoid](https://mxnet.apache.org/versions/1.8.0/api/python/docs/api/ndarray/ndarray.html#mxnet.ndarray.hard_sigmoid)

**Comment by igorsafo** [comment]
Hi @zamir1001 ,
Thank you for your request!

As I understand hardsigmoid can be implemented using already implemented eltwise algorithms in oneDNN:
```cpp
tmp = eltwise_linear(x, alpha, beta);
y = eltwise_bounded_relu(tmp, 1.0f);
```
Does this solve you problem?

**Comment by zamir1001** [comment]
Hi @igorsafo,

For the moment I have a custom implementation that's supports training.

For the forward pass I could use
```
tmp = eltwise_linear(x, alpha, beta);
y = eltwise_bounded_relu(tmp, 1.0f);
```

but how do I deal with the backward pass?

**Comment by igorsafo** [comment]
For backward pass you could use the same 2 primitives:
```
// dy/x = 0 < alpha * x + beta < 1
// dy/x = 0 (otherwise)
diff_tmp = eltwise_bounded_relu(backward_data, tmp, dy, 1.0f);
// dy/x = alpha
dx = eltwise_linear(backward_data, x, diff_tmp, alpha, beta);
```

**Comment by zamir1001** [comment]
Hi @igorsafo,

that's indeed a perfect solution.
I will implement the hard logistic like this.
There's no need for a separate dnnl_elementwise_hardlogistic function.

thanks

**Comment by igorsafo** [comment]
Correct, "native" `elementwise_hardlogistic` would require only `src` data from forward pass, while combination of `eltwise_bounded_relu` and `eltwise_linear` requires `src` and `tmp` buffers from forward pass.

**Comment by zamir1001** [comment]
I switched to dnnl_elementwise_logistic. I think the performance benefit of a "native" elementwise_hardlogistic is probably not that big when using an Intel64 cpu. 

thanks

**Comment by TaoLv** [comment]
JFYI, hard sigmoid algorithm has been supported on master via https://github.com/oneapi-src/oneDNN/commit/e703d23ccad127cd1e4f52330313dd6f5190ca01 . Thanks to @dzarukin .

**Comment by zamir1001** [comment]
Hi,

I've noticed the change! I've already adapted my code.
Thanks very much.
